libraries/matrix/impl/src/main/kotlin/io/element/android/libraries/matrix/impl/encryption/RustEncryptionService.kt

```diff
@@ -88,22 +97,37 @@ internal class RustEncryptionService(
 
     override val enableRecoveryProgressStateFlow: MutableStateFlow<EnableRecoveryProgress> = MutableStateFlow(EnableRecoveryProgress.Starting)
 
+    /**
+     * Check if the session is the last session every 5 seconds.
+     * TODO This is a temporary workaround, when we will have a way to observe
+     * the sessions, this code will have to be updated.
+     */
+    override val isLastDevice: StateFlow<Boolean> = flow {
+        while (currentCoroutineContext().isActive) {
+            val result = isLastDevice().getOrDefault(false)
+            emit(result)
+            delay(5_000)
+        }
+    }
+        .stateIn(sessionCoroutineScope, SharingStarted.Eagerly, false)
+
     fun start() {
-        service.backupStateListener(object : BackupStateListener {
+        backupStateListenerTaskHandle = service.backupStateListener(object : BackupStateListener {
```

Comment:
Use `mxCallbackFlow` instead, so we can remove this `start` function. Maybe create extension methods on `Encryption` so you can reuse them.

```
private val backupStateFlow = mxCallbackFlow {
        val listener = object : BackupStateListener {
            override fun onUpdate(status: RustBackupState) {
                trySend(backupStateMapper.map(status))
            }
        }
        service.backupStateListener(listener)
    }.stateIn(sessionCoroutineScope, SharingStarted.Eagerly, service.backupState().let(backupStateMapper::map))
```

```
 private val recoveryStateFlow = mxCallbackFlow {
        val listener = object : RecoveryStateListener {
            override fun onUpdate(status: org.matrix.rustcomponents.sdk.RecoveryState) {
                trySend(recoveryStateMapper.map(status))
            }
        }
        service.recoveryStateListener(listener)
    }.stateIn(sessionCoroutineScope, SharingStarted.Eagerly, service.recoveryState().let(recoveryStateMapper::map))
```